### DISCUSSION TITLE
Cannot verify signature

### DISCUSSION THREAD

**Discussion by beegentoo** [panva/jose#356] [Q&A]
Hello everybody,

**EDIT: Problem partially solved -- see bottom paragraph**

I am experiencing problems verifying a signature using a public-private-keypair in nodejs. Note: Promise Handling is quite verbose due to my futile attempts to examine this problem.

I create a keypair using `let keyPair = jose.generateKeypair('RS256')`. The public key is then exported via `jose.exportSPKI(keyPair.publicKey)`. This exported public key is made available via HTTP and looks like this:
```
-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAvaOQ2/hgX0PpuVFDuTpK
20gSyfnkYV6wK5/3ujpbAa1Vp9a6hLFxqQjcXhwiBvyh4FYWMGy00OnWkUL7Vu8g
s5aip7W8FdLt0NczVbjvheDKmqP8YhOswjhTQbG1NuQtcny6zwfYZpoONm9ryy2i
QtsTQt9jn5Q1TXZF5tlo9bArVSfSdfVrsLvULx/zUaxz6vK9satY3mGOw0kiD238
+PonUYVUOz9xPdgbcZCFyzdKJ2PwXh/5bEP6aWbGr2sOYLR5Q6a/raQ9lllktRHO
ScRcG/XltQC6swIxB0tBOi4u3l16vlGWwHcxmpWQdCqlZxlFJRhgXt2TrywdulUj
lwIDAQAB
-----END PUBLIC KEY-----

```
I then create a JWT and sign it with the private key:
```JavaScript
return new Promise((resolve, reject) => {
    let payload = {
        user: authContainer.sessionData.user,
        groups: authContainer.sessionData.groups
    }
    var jwt = new jose.SignJWT(payload);

    jwt.setProtectedHeader({alg: 'RS256'});
    jwt.setIssuedAt();
    jwt.setIssuer('DeputyAuth');
    jwt.setAudience('Microservices');
    jwt.setExpirationTime('1h');
    jwt.sign(keyPair.privateKey).then(jwt => {
        authContainer.jwt = jwt; // the jwt is added to an "envelope"-Container since more data is passed back to the caller.
        resolve(authContainer);
    }).catch(error => {
        logger.error("Error signing jwt");
        logger.error(error);
        reject(authContainer);
    });
});
```
No probems so far. I get a "readable" JWT. However, when verifying the signature in a consuming application, the signature verification fails.

The public key gets fetched via HTTP (as mentioned above) and is then imported using `jose.importSPKI()` and is then used to verify the JWT which is being delivered via the Authorization-Header (the "Bearer"-Prefix is cut before validation):
```JavaScript
jose.compactVerify(plainToken, pubkey).then(result => {
    logger.info(result);
}).catch(error => {
    logger.error(error);
});
```
This leads to the following log-entry:
```JSON
{
  "code":"ERR_JWS_SIGNATURE_VERIFICATION_FAILED",
  "level":"error",
  "message":"signature verification failed",
  "module":"Main",
  "name":"JWSSignatureVerificationFailed",
  "timestamp":"2022-02-03T09:10:08.246Z"
}
```

I am a bit lost with this error since I have no idea where and how to further examine. Maybe there are some well known common caveats I don't know yet but someone else does.

Thanks in advance for any hints!

**Partial Solution:** It seems Key-Pairs using RS256 or PS256 don't work. Using an EC256 pair signature verification works.

Ben

**Comment by panva** [comment]
Please provide the steps to reproduce as a simple script we can run and details about your javascript runtime.

**Reply by beegentoo** [reply]
This would be the demo to demonstrate working/non-working -- unfortunately here, both algorithms work :-/

```JavaScript
const jose = require('jose');

const ALGO_BROKEN = 'RS256';
const ALGO_WORKING = 'ES256';

async function errorDemo(algo) {
    console.log(`Demonstration using algo ${algo}`);

    // Prepare key pair
    try {
        var keyPair = await jose.generateKeyPair(algo);
        var pubkey = await jose.exportSPKI(keyPair.publicKey);
        console.log(pubkey);
    } catch(error) {
        console.error(`Error creating/exporting key-pairs for algo ${algo}`);
    }

    // Create token, sign with private key
    let payload = {
        user: 'JohnDoe',
        groups: ['Group A', 'Group B']
    }

    let jwt = new jose.SignJWT(payload);
    jwt.setProtectedHeader({alg: algo});
    jwt.setIssuedAt();
    jwt.setIssuer('DeputyAuth');
    jwt.setAudience('Microservices');
    jwt.setExpirationTime('1h');

    try {
        var signedToken = await jwt.sign(keyPair.privateKey);
    } catch(error) {
        console.error(`Error signing token for algo ${algo}`);
    }

    //Verify the token using the previously exported public key -- This would normally be fetched via HTTP
    try {
        var importedPubkey = await jose.importSPKI(pubkey);
        var extractedPayload = await jose.compactVerify(signedToken, importedPubkey);
        console.log(`Token is valid for algo ${algo} : `+ extractedPayload.payload.toString());
    } catch(error) {
        console.error(`Token is invalid for algo ${algo} -- error: ${error}`);
    }
}

async function showDemos() {
    await errorDemo(ALGO_BROKEN);
    await errorDemo(ALGO_WORKING);
}

showDemos();
```
About the runtime:

- Nodejs 12.22.10
- jose `"jose": "^4.4.0"`
- Running on Ubuntu 20.04.3 LTS

**Reply by panva** [reply]
🤷‍♂️ well, is it not possible you are just not verifying with the right key? The error would be `ERR_JWS_SIGNATURE_VERIFICATION_FAILED` in such case.

Also, if you're using JWT with expiration you should be using jwtVerify instead of compactVerify.

**Reply by beegentoo** [reply]
I get exactly this error but I manually checked while debugging that the public key being used is the one belonging to the private key being used for signing -- but of course I will check again. 

And regarding the verification method: I know but since the error only mentioned the signature I am just checking this for the demo-case here and not issuer and audience.
